src/context.rs

```diff
@@ -39,10 +39,18 @@ pub fn get_buffer_words(buf: &Buffer) -> Vec<(usize, usize)> {
     res
 }
 
+/// The key bindings to use.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum KeyBindings {
+    Vi,
+    Emacs,
+}
+
 pub struct Context {
     pub history: History,
     pub completer: Option<Box<Completer>>,
     pub word_fn: Box<Fn(&Buffer) -> Vec<(usize, usize)>>,
+    pub key_bindings: KeyBindings,
```

Comment:
Could also add a `State` item and a `state()` method to `KeyMap` that will return the state. Then when a new `KeyMap` is created the state could be passed in. Seems a little complicated. Could also extract the bits of state we need (clipboard, undo buffers, etc) and treat them like history from `Context`. There could still be some state specific to some keymap that needs to be preserved though.